### PR TITLE
[framework]Update hook to ensure check_dut_asic_type runs before any other module-level fixtures

### PR DESCRIPTION
**Pull request by bingwang-ms**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

#### What is the motivation for this PR?
Fixture check_dut_asic_type is set to function level, as a result, it always runs after all module-level and function-level fixtures in default settings of pytest. 
However, some module or function level fixtures are unnecessary to run if the case is to be skipped, such as
https://github.com/Azure/sonic-mgmt/blob/438cc1f56672c1ff0036bedf4a69a8b156c81efe/tests/platform_tests/broadcom/test_ser.py#L50-L56
and
https://github.com/Azure/sonic-mgmt/blob/438cc1f56672c1ff0036bedf4a69a8b156c81efe/tests/vxlan/test_vnet_route_leak.py#L37-L58
These fixtures are either time consuming or leave the DUT in unstable status.
This commit update the hook for custom fixture to ensure check_dut_asic_type runs before any other module-level and function-level fixtures . 
#### How did you do it?
Update **pytest_collection_modifyitems** hook. The check_dut_asic_type fixture is inserted into fixturelist before all module-level fixture, so that those fixtures behind it no longer runs if the case is to be skipped.
#### How did you verify/test it?
I verified it with a sample script
```
  import pytest
  import logging

  pytestmark = [
      pytest.mark.topology('any'),
      pytest.mark.asic('mellanox')
  ]

  @pytest.fixture(scope="module", autouse=True)
  def fixture_for_module():
      logging.warn("module level fixture")

  class TestClass:
      def test_class_1(self):
          pass
      def test_class_2(self):
          pass

  def test_should_skip(fixture_for_module):
      pass

  @pytest.mark.asic('broadcom')
  def test_rewrite_marker():
      pass

  @pytest.mark.asic()
  def test_should_run():
      pass
```
And the fixture and hook work as expected.
```
collected 5 items

platform_tests/demo/test_asic.py::TestClass::test_class_1 SKIPPED                                                                                                                               [ 20%]
platform_tests/demo/test_asic.py::TestClass::test_class_2 SKIPPED                                                                                                                               [ 40%]
platform_tests/demo/test_asic.py::test_should_skip SKIPPED                                                                                                                                      [ 60%]
platform_tests/demo/test_asic.py::test_rewrite_marker
------------------------------------------------------------------------------------------- live log setup --------------------------------------------------------------------------------------------
15:10:14 WARNING test_asic.py:fixture_for_module:11: module level fixture
PASSED                                                                                                                                                                                          [ 80%]
platform_tests/demo/test_asic.py::test_should_run PASSED                                                                                                                                        [100%]

------------------------------------------------------------------ generated xml file: /data/Networking-acs-sonic-mgmt/tests/tr.xml -------------------------------------------------------------------
================================================================================= 2 passed, 3 skipped in 7.74 seconds =================================================================================
```
#### Any platform specific information?
No.
#### Supported testbed topology if it's a new test case?
N/A
### Documentation 

No

### CODE REVIEW

File: tests/common/plugins/custom_fixtures/__init__.py [hunk 1]
```diff
@@ -6,5 +6,11 @@ def pytest_collection_modifyitems(items):
...
         asic_marks = [mark for mark in item.iter_markers(name="asic")]
         if asic_marks:
-            item.fixturenames.append("check_dut_asic_type")
+            for pos, name in enumerate(item.fixturenames):
+                try:
+                    if item._fixtureinfo.name2fixturedefs[name][0].scope == 'module':
+                        break
+                except:
+                    pos = len(item.fixturenames)
+            item.fixturenames.insert(pos, "check_dut_asic_type")
```
Comment (neethajohn) [inline tests/common/plugins/custom_fixtures/__init__.py:15]: if this loop executes to completion, then this fixture gets inserted as the second last item. Is that what is intended?

Comment (bingwang-ms) [reply tests/common/plugins/custom_fixtures/__init__.py:15]: Thanks Neetha. It's a bug. I'll post an update.

Comment (bingwang-ms) [reply tests/common/plugins/custom_fixtures/__init__.py:17]: Updated. Please help to re-check. Thanks @neethajohn

**Comment by wangxin** [comment]
In the PR description, what is the command line for generating the test log?

**Comment by bingwang-ms** [comment]
> In the PR description, what pytest command line arguments were used for running the script for test? On which platform was the test script executed?

The Platform I ran the sample script is broadcom, and there is nothing special in command line
```
py.test --inventory $INVT --host-pattern $DUT_NAME --module-path $MPATH --testbed $TESTBED_NAME --testbed_file $TESTBED_FILE --junit-xml=tr.xml --log-cli-level ${LOG_LEVEL} --collect_techsupport=False --topology='t0,any,util' platform_tests/demo/test_asic.py
```

**Review by wangxin — APPROVED** [review]

**Pull request merged**